### ISSUE TITLE
Key in index is not an unstructured

### ISSUE THREAD

**Issue by alexec** [argoproj/argo-workflows#4308]
## Summary 

```
controller | time="2020-10-16T13:55:39-07:00" level=warning msg="Key in index is not an unstructured" key="{argo/awesome-bear-hg6qr 0xc0006b0090}"
 controller | E1016 13:55:39.832145   21218 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x23b3dc0), concrete:(*runtime._type)(0x2465b40), asserted:(*runtime._type)(0x25f4f40), missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *unstructured.Unstructured)
 controller | goroutine 182 [running]:
 controller | k8s.io/apimachinery/pkg/util/runtime.logPanic(0x23f7160, 0xc000bac660)
 controller | 	/Users/acollins8/go/pkg/mod/k8s.io/apimachinery@v0.17.8/pkg/util/runtime/runtime.go:74 +0xa3
 controller | k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
 controller | 	/Users/acollins8/go/pkg/mod/k8s.io/apimachinery@v0.17.8/pkg/util/runtime/runtime.go:48 +0x82
 controller | panic(0x23f7160, 0xc000bac660)
 controller | 	/usr/local/Cellar/go@1.13/1.13.15/libexec/src/runtime/panic.go:679 +0x1b2
 controller | github.com/argoproj/argo/workflow/controller.(*WorkflowController).addWorkflowInformerHandlers.func9(0x2465b40, 0xc000db3ca0)
```

## Diagnostics

What Kubernetes provider are you using? 

What version of Argo Workflows are you running? 

```yaml
Paste a workflow that reproduces the bug, including status:
kubectl get wf -o yaml ${workflow} 
```

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
@simster7 could I please ask you to own this? It is an edge-case that only happens (I think) when a watch returns an object that was deleted.

**Comment by simster7** [comment]
Absolutely

**Comment by alexec** [comment]
It maybe not fixable - I'll leave it to you. I think this is the code you want.

```golang
			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
			if err == nil {
	//
			}
```

**Comment by simster7** [comment]
Seems like that is the correct approach

**Comment by alexec** [comment]
Fixed in  #4307.
